### ISSUE TITLE
Enhancement: add renderer.outputEncoding

### ISSUE THREAD

**Issue by WestLangley** [mrdoob/three.js#11337]
As discussed in #11332

1. `renderer.gammaInput` needs to be deprecated. We now support `texture.encoding` instead.

2. renderer.gammaOutput needs to be deprecated and replaced with `renderer.outputEncoding` when rendering to the screen. We already support `renderTarget.texture.encoding` when rendering to a render target.

3. WebGLPrograms.getProgramCode()  needs to be updated to respond to changes in `renderer.outputEncoding`.  Setting `material.needsUpdate = true` will then force the change.

4. Legacy warnings need to be added, and the examples need to be updated to accommodate this change.

**Comment by WestLangley** [comment]
Actually, before we go down that path, perhaps we should decide on decoupling "encoding" into _color space_ and _encoding format_,  as [suggested](https://github.com/mrdoob/three.js/pull/11076#issuecomment-293570510) by @bhouston.

>If we split color space conversion from the encodings, we could do the above. It would be more flexible and clear. We could also support LUTs in the color space conversions as well.

>And then each texture and render target should have both a ColorSpace (Linear, sRGB, AdobeRGB, Gamma20, Gamma22) and a Encoding (None, RGBE, RGBM, RGBD, LogLUV, CMYK) parameter.

This is something I would support.

**Comment by bhouston** [comment]
> Actually, before we go down that path, perhaps we should decide on decoupling "encoding" into color space and encoding format, as suggested by @bhouston.

Separating colorSpace and encoding is not that hard.  I would just copy the code used to inject functions for encoding into the glsl and modify it to inject colorSpace transforms.  And then split the enums into a color space list and an encoding list.  I guess I am saying just use the existing design for encodings, but duplicate it again for color spaces.

I think that is sort of the easy, obvious solution.

**Comment by bhouston** [comment]
The injectors are pretty simple - they just create specially named functions that map onto the chosen encoding function: 

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L12

I use the same injection idea for tonemapping: getToneMappingFunctio

**Comment by bhouston** [comment]
Hmm... I wonder if tone mapping could be combined with color spaces?  They are sort of the same thing theoretically -- a color transform.  And they sort of happen at the same time in the renderer pipeline (before the encoding.)

And then moved from a global on WebGLRenderer to something that is texture specific (except we would use the global tone mapping operator when one doesn't have a render target, in the same fashion that WebGLRenderer.outputEncoding would be used.)

I have run into issues with the global tone mapping operator -- and as it is currently designed it is annoying to disable and re-enable.  And I think a per texture colorspace/tonemapper may be more useful, and impose no limitations that I can think of.

Is there precedent for considering tonemapping a form of color space conversion?  I wonder because it would justify combing them.

**Comment by bhouston** [comment]
I think that the order maybe:

toneMapping (gamut reduction)
colorSpace (gamut transforms, assumes inputs are between 0-1)
encoding (assumes inputs are between 0-1)

I guess you may want to use a Reinhart tone map operator may or may not want sRGB output -- thus that argues for a separation of toneMapping from colorSpace.

Another complicating is that HDR TVs and Monitors are coming and they break the assumption that colorSpaces have values between 0 and 1, but they still use gamma curves I believe: https://en.wikipedia.org/wiki/Rec._2020

**Comment by WestLangley** [comment]
To summarize:

    toneMapping ( to LDR )
    colorSpace ( Linear, sRGB, AdobeRGB, Gamma20, Gamma22, Grayscale, CustomLookup )
    encoding ( RGBA, RGBE, RGBM, RGBD, LogLUV, CMYK )

This means all textures and render-target textures have the following two properties:
```
texture.colorSpace = THREE.sRGBColorSpace; // for example
texture.encoding = THREE.NoEncoding; // is THREE.RGBAEncoding a better term ???
```
So instead of adding `renderer.outputEncoding`, we would add:
```
renderer.colorSpace = THREE.sRGBColorSpace; // THREE.LinearColorSpace, etc..
// renderer.encoding = THREE.RGBAEncoding; // I guess that is not needed
```
>that argues for a separation of toneMapping from colorSpace.

Yes.

**Comment by bhouston** [comment]
Yeah.  Just to be clear I see this as an eventual end point:

```
texture.toneMapping
texture.colorSpace
texture.encoding

// and for the case of the null renderTarget, rendering to screen (which is sort of an annoying exception):
renderer.colorSpace
renderer.toneMapping
renderer.encoding
```

I guess putting these colorSpace/toneMapping/encodings on textures/renderTargets could get weird when one reuses them in the EffectComposer -- you need to be careful what you set because we recycle readBuffer, writeBuffer as we step through the composer.  I guess we already have that problem, so maybe it makes no real issue right now.

**Comment by bhouston** [comment]
Actually you are right, no encoding needed for renderer.  It should always be RGBAEncoding.

**Comment by WestLangley** [comment]
`texture.toneMapping` does not sound right to me...

I am now thinking we need to separate further:
```
toneMapping ( HDR to LDR )
colorMapping ( to Grayscale, to CustomLookupTable )
colorSpace ( Linear, sRGB, AdobeRGB, Gamma20, Gamma22 )
encoding ( RGBA, RGBE, RGBM, RGBD, LogLUV, CMYK )
```

Tone mapping in this case is really a property of a render target. The renderer is a stand-in for the default render target:
```
renderer.toneMapping (when rendering to the screen)
renderTarget.toneMapping
```
Likewise, color mapping: it is essentially a post-processing effect, but applied as-we-go.
```
renderer.colorMapping (when rendering to the screen)
renderTarget.colorMapping
```
Color space is a property of a texture, render target texture, and renderer output.
```
texture.colorSpace
renderTarget.texture.colorSpace
renderer.colorSpace
```
Encoding is a property of a texture and a render target texture.
```
texture.encoding
renderTarget.texture.encoding
```
So, to summarize
```
renderer.toneMapping
renderer.colorMapping
renderer.colorSpace
//renderer.encoding - not needed; always RGBA

renderTarget.toneMapping
renderTarget.colorMapping

texture.colorSpace
texture.encoding
```
That's the way I see it... for now. :)

**Comment by bhouston** [comment]
Well, we could view this as a list of inline post processing transforms.  Encoding we run through the set forward with forward transforms, decoding we go backwards with inverse transforms.

Such an interface would look like:

```
texture.texelOps.add( THREE.FilmicToneMapping, { exposure: exposure, whitePoint: whitePoint } );
texture.texelOps.add( THREE.sRGBColorSpace );
texture.texelOps.add( THREE.RGBMEncoding, { multiplier: 16 } );
```

This reduces it to a single list of transforms/operators, and you can have an arbitrary number.  Implementation would be simpler and it would be more flexible going forward.  Maybe each operator could be a class that contains its own code injection method so it is better encapsulated and infinitely extensible by others -- rather than hard coded enums on THREE.

I've never seen this design paradigm before.  That doesn't mean it is wrong, but it makes me not super confident in pushing it.  But if others agree....

**Comment by bhouston** [comment]
The texel op class would look like this -- this is the default that doesn't do anything.  You would override it with better data.
```
class TexelOp {
   fowardOp: function( name, parameters ) {
      return "vec4 " + name + "( vec4 input ) { return input; }";
   }
   backwardOp: function( name, parameters ) {
      return "vec4 " + name + "( vec4 input ) { return input; }";
   }
}
```

**Comment by mrdoob** [comment]
@bhouston I like that approach!

**Comment by donmccurdy** [comment]
It may help avoid surprises for new users if we changed the default `renderer.gammaOutput=true`, or whatever the equivalent `outputEncoding` or `texelOp` syntax ends up being... Is it worthwhile to do that, and deprecate `gammaInput`, before we implement a more flexible interface here?

GLTFLoader automatically sets `material.map.encoding = THREE.sRGBEncoding` (required by glTF, and likely to be generally correct for other formats) and assumes the user has set `gammaOutput=true`. Frequently, [users find this confusing](https://twitter.com/jorgefuentesnet/status/1001143746786885632).

**Comment by bhouston** [comment]
I believe that renderer.gammaOutput should be deprecated and instead use renderer.outputEncoding instead.

**Comment by WestLangley** [comment]
Right. See point 2 in original post.

**Comment by donmccurdy** [comment]
If we're renaming it anyway, why not `renderer.colorSpace` or `renderer.outputColorSpace` per https://github.com/mrdoob/three.js/issues/11337#issuecomment-301599802? It seems like encoding is out of place, and `toneMapping` can be added later.

**Comment by bhouston** [comment]
It was my mistake to combine both colorSpace and encodings when implementing encodings.  They should have been separate.

BTW I am looking at a lot at 3D LUTs recently for colorMappings, they can produce pretty beautiful results.

**Comment by WestLangley** [comment]
>BTW I am looking at a lot at 3D LUTs recently for colorMappings, they can produce pretty beautiful results.

links? :-)

**Comment by bhouston** [comment]
Uses color mapping:
https://wizgrav.github.io/aframe-effects/
https://github.com/wizgrav/aframe-effects/tree/master/components/colors

Originally based on this demo, which has disappeared unfortunately: https://twitter.com/mattdesl/status/712043209652903936?lang=en

Color mapping is also known as "color grading" in the film industry.  Good introductory video: https://www.youtube.com/watch?v=4sblEu4x5ug

Without 3D LUT support in Three.JS we can not get the "cinematic look" without doing it the hard way using lights and materials.  When really you can just achieve a gritty and emotional look via a 3D LUT on top of a normal looking scene.

**Comment by bhouston** [comment]
@zomboh Can you mix different gamma with sRGB and AdobeRGB?  My understanding is that you can not.  Gamma2.2 is sort of a poor man's sRGB?

From the sRGB wikipedia article you can see that sRGB includes a gamma curve: https://en.wikipedia.org/wiki/SRGB

"sRGB uses the ITU-R BT.709 primaries, the same as in studio monitors and HDTV,[2] a transfer function (gamma curve) typical of CRTs, and a viewing environment designed to match typical home and office viewing conditions.  This specification allowed sRGB to be directly displayed on typical CRT monitors of the time, which greatly aided its acceptance."

So I believe that sRGB should not be combined with another gamma.  You will apply double gamma correction.

In the Adobe RGB wikipedia article: https://en.wikipedia.org/wiki/Adobe_RGB_color_space

"As with sRGB, the RGB component values in Adobe RGB (1998) are not proportional to the luminances. Rather, a gamma of 2.2 is assumed, without the linear segment near zero that is present in sRGB. The precise gamma value is 563/256, or 2.19921875. In coverage of the CIE 1931 color space the Adobe RGB (1998) color space covers 52.1%"

Thus I believe that AdobeRGB, sRGB and Gamma2 and Gamma22 are color spaces and they should be mutually exclusive.

**Comment by bhouston** [comment]
I can see us adding ACES color standards are based in the HDTV/4K standard Rec._2020 which itself defines a gamma correction function to its linear colors:

https://en.wikipedia.org/wiki/Rec._2020#Transfer_characteristics

I am still a bit confused as to whether Rec 2020 is natively linear as stated in the ACES article or whether it is gamma corrected natively.  But either way, I believe that it is mutually exclusive with the other transforms.  You get to pick one, not a mix and match approach.

**Comment by zomboh** [comment]
@bhouston you're right! I stand corrected. I think I mixed up the gamma with the white point since you can target both to an arbitrary value in a calibration, but in the case of a color space there is a fixed gamma value, indeed. Thanks for the clarification!

**Comment by donmccurdy** [comment]
I'm strongly considering a change to how A-Frame (http://aframe.io) handles color, based on the [two "workflows" in Unity](https://docs.unity3d.com/Manual/LinearRendering-LinearOrGammaWorkflow.html).  Essentially:

- `workflow=gamma` — same as current three.js defaults, except that `texture.encoding` is ignored. Colors and textures are used as they are stored, even though shader calculations treat inputs as if they were linear space. Output is _not_ gamma-corrected.
- `workflow=linear` — all colors inputs are assumed to sRGB, and converted to linear when passed to the GPU, textures are interpreted according to `texture.encoding`, and converted to linear in the shader. Output is gamma-corrected (unless you opt out for post-processing).

Mentioning it here because I think this syntax may be easier for users to understand than input/output gamma, encoding, and colorspace. In either workflow, input and output conversions are managed for you. Input `material.color.setHex( 0x4285f4 )` will look the same as `#4285f4` in CSS, assuming full lighting. That is not true of three.js when `gammaOutput: true`.

It could be worth considering a similar "workflow" option in three.js. Even if we can't simplify it down to a single setting like that, documenting the two workflows and the specific settings needed to achieve them would help developers.

**Comment by WestLangley** [comment]
@donmccurdy I am inclined to agree with you.

However, in your Unity link, I was surprised to read:

>These inputs are then passed to the Shader, with lighting calculations taking place in linear space as they normally do.

That sounds different than "treated as if they were in linear space"...

I never liked the fact that colors in three.js are assumed to be in linear space, as this conflicts with CSS, where they are in sRGB space.

**Comment by looeee** [comment]
> Even if we can't simplify it down to a single setting like that, documenting the two workflows and the specific settings needed to achieve them would help developers.

I agree - reading discussions here and on the forum, it seems like color spaces and texture encoding are a major point of confusion for users at the moment, beginners and experts alike.

Clearly documented workflows would be a good starting point for improvements to the API.

**Comment by donmccurdy** [comment]
> > These inputs are then passed to the Shader, with lighting calculations taking place in linear space as they normally do.
> 
> That sounds different than "treated as if they were in linear space"...

I don't follow what you mean here, or for that matter what the Unity docs refer to with "as they normally do." The Unity quote is from the "linear" workflow, whereas your quote from my post is describing the "gamma" workflow. But I think the gist (in the preferred linear workflow) would be that when the user selects a color in the Unity Editor, gamma correction is removed when creating the GPU constants in the linear workflow. In effect that shields the user from having to worry about all of this. 😅

**Comment by bhouston** [comment]
I am not sure we should support this weird gamma mode.  It is just simply wrong and I do not see a reason to support it except that Unity has also implemented a "wrong" mode.

The future is purely linear workflows.  Now I am open to improving its usefulness and to remove surprises but I think encouraging a clearly wrong mode is not the answer.

**Comment by bhouston** [comment]
> I agree - reading discussions here and on the forum, it seems like color spaces and texture encoding are a major point of confusion for users at the moment, beginners and experts alike.

I think this can be solved with documentation and maybe a chart or something.  We should not muddy the waters by having a mode that ignores texture.encoding, that is just wrong.  I would only suggest something like that for backwards compatibility sake.

**Comment by bhouston** [comment]
> I never liked the fact that colors in three.js are assumed to be in linear space, as this conflicts with CSS, where they are in sRGB space.

We should maybe have a table that says the encoding of each parameter.  I am okay with changing the color space of the color specification variables to what is most standard in other tools (Unity, UE4, Maya, 3DS Max, V-Ray, etc.)

**Comment by donmccurdy** [comment]
I agree that the gamma mode has no benefit other than backwards compatibility, but unfortunately that's a major sticking point... If we were to introduce a gamma mode to three.js, probably we should keep it _exactly_ the same as the current defaults, rather than ignoring `texture.encoding` as Unity does.

I'd be even happier with simply going entirely to a linear workflow — I agree linear is the preferred direction. If we do the following, existing scenes would have a (hopefully) smooth update:

- (1) default to remove gamma correction from `THREE.Color` at GPU upload
- (2) default to apply gammaOutput=true (or whatever we rename it to)
- (3) default to consider color maps sRGB

Unfortunately (3) is hard. Most users are not setting `texture.encoding` today, and most loaders do not set `texture.encoding` on their textures either, so everything uses the default `THREE.LinearEncoding`. We'd have to ignore that setting, or users would need to start setting encoding on all of their textures when they update. A possible workaround would be to introduce...

```
texture.encoding = THREE.DefaultEncoding
```

... which leaves the renderer free to infer an encoding from the workflow, format, and map slot. I hope that is a reasonable API. We'd probably want to do 1-3 all in a single release.

**Comment by bhouston** [comment]
> (1) default to remove gamma correction from THREE.Color at GPU upload

Remove gamma or remove sRGB.  Those are slightly different.  I am not sure this needs to be done but it is a choice.

> (2) default to apply gammaOutput=true (or whatever we rename it to)

gammaOutput=true is there for backwards compatibility.  We really should have renderer.outputColorSpace=sRGB instead.

> (3) default to consider color maps sRGB

This is acceptable in the case of LDR maps, where as HDR maps should be assumed to be linear.

texture.colorSpace = sRGB or Linear
texture.texelEncoding = Raw, RGBM16, RGBD, RGBM8, LogLuv

colorSpace and texelEncoding should be separated.  They are currently conflated and this leads to confusion.

**Comment by bhouston** [comment]
Call it AutoColorSpace rather than default as auto implies it switches
smartly based on context.  I am fine changing the color space for material
parameters.

**Comment by MasterJames** [comment]
I thought to stress my preference for Filmic.
This video is comprehensive info about having high dynamic range.
https://youtu.be/m9AT7H4GGrA
I realize realtime GPU may not be optimal performance (FPS) for high bitdepth but as GL moves forward with realism. As the future moves forward in Blender with Eevee I felt those building ThreeJS might appreciate a reminder of the potential and the rising bar.
https://youtu.be/gy4E9nc5m0E

Anyway I guess for clarity the point is you cant get photoreal (realtime) interactive without high dynamic range. If content hdr light maps textures models etc. are best coming from Blender, then using the 3rd party filmic encoding is the right way in that camp.

Anyway I'm hopeful that working in Blender with Eevee will translate well to three.js via web, and just want people to be mindful of those ideals.

**Comment by bhouston** [comment]
The way hdr is implemented in three.js means it costs almost nothing
because it uses rgba8 for storage.  I agree that filmic tone mapping on by
default is best.

**Comment by WestLangley** [comment]
>I agree that the gamma mode has no benefit other than backwards compatibility

https://github.com/mrdoob/three.js/issues/15100#issuecomment-431818277 may be a counter-argument...

**Comment by donmccurdy** [comment]
Here are early results from A-Frame scenes, switching to a linear workflow. Changes:

(1) color inputs are assumed sRGB
(2) color textures are assumed sRGB
(3) gammaOutput on by default

|#| gamma | linear |
|--|--------|-------|
|1 |![screen shot 2018-12-03 at 11 59 55 pm](https://user-images.githubusercontent.com/1848368/49427370-02259400-f758-11e8-8203-45c34031dc1f.png)| ![screen shot 2018-12-04 at 12 00 15 am](https://user-images.githubusercontent.com/1848368/49427374-03ef5780-f758-11e8-9a5e-9b9cd31aadd7.png)|
|2 | ![screen shot 2018-12-03 at 10 38 49 pm](https://user-images.githubusercontent.com/1848368/49423651-05675280-f74d-11e8-857f-056ee5800152.png) | ![screen shot 2018-12-03 at 10 38 50 pm](https://user-images.githubusercontent.com/1848368/49423652-06987f80-f74d-11e8-8094-033be1521a22.png) |
|3 | ![screen shot 2018-12-03 at 11 59 03 pm](https://user-images.githubusercontent.com/1848368/49427390-11a4dd00-f758-11e8-8919-23dc52e7b9c3.png) | ![screen shot 2018-12-03 at 11 59 09 pm](https://user-images.githubusercontent.com/1848368/49427394-136ea080-f758-11e8-90f3-8a66cfce8661.png) |
| 4 | ![screen shot 2018-12-03 at 11 59 28 pm](https://user-images.githubusercontent.com/1848368/49427407-19648180-f758-11e8-9cd2-10276f372e12.png) | ![screen shot 2018-12-03 at 11 59 29 pm](https://user-images.githubusercontent.com/1848368/49427411-1b2e4500-f758-11e8-9b9f-f4f48e6c3b0e.png)| 
|5 |![screen shot 2018-12-04 at 12 02 13 am](https://user-images.githubusercontent.com/1848368/49427422-21242600-f758-11e8-80f4-90a1dc17ff7c.png)|![screen shot 2018-12-04 at 12 02 12 am](https://user-images.githubusercontent.com/1848368/49427423-22555300-f758-11e8-90e4-9d1fca8ab952.png) |
|6 | ![screen shot 2018-12-03 at 10 42 43 pm](https://user-images.githubusercontent.com/1848368/49423728-3a73a500-f74d-11e8-8986-75183e34443d.png) | ![screen shot 2018-12-03 at 10 42 44 pm](https://user-images.githubusercontent.com/1848368/49423734-3cd5ff00-f74d-11e8-8adc-653fc3b9fdf0.png) |

The overall trend, at least for these examples, appears to be that shading is slightly less pronounced. The differences are definitely more acceptable for us than changing default values for any of those settings alone would have been.

**Comment by bhouston** [comment]
Just remember one of the reasons for the slight difference is that sRGB to linear and then linear to gamma 2.0 won't restore the inputs.  The last encode should be to sRGB instead.

**Comment by donmccurdy** [comment]
Here's another comparison displaying possible corrections to threejs color management. In the links below (and [this branch](https://github.com/mrdoob/three.js/compare/dev...donmccurdy:feat-color-management?w=1)) I've applied some significant changes:

- remove `gammaInput`
- change default `texture.encoding` to `null`
- add `Color.colorManagement`, default=true, which causes user hexadecimal and HSL color inputs to be treated as sRGB. The getter/setter methods do this conversion, so `color.toArray()` and `color.r/g/b` are linear.
- add `renderer.colorManagement`, default=true, which causes color maps to be treated as sRGB, unless otherwise specified.
- add `renderer.linearOutput`, default=false, which disables linear->sRGB encoding of output (e.g. for use with postprocessing).

I'm not confident that's the right user API, but the end result should be correct. The effect is noticeable – especially so on examples that weren't configured quite correctly before, such as enabling `gammaOutput=true` but not marking color textures as sRGB. I haven't attempted to fix any of the examples in the comparisons below.

[Demo](https://rawcdn.githack.com/donmccurdy/three.js/15601937339c574f536d3cb96297d4d29f62b95f/examples/index.html) (synced to r101)

| gamma (before) | linear (after) |
|----|----|
| ![screen shot 2019-01-31 at 11 23 06 pm](https://user-images.githubusercontent.com/1848368/52108686-9a50c580-25af-11e9-8484-01a309049e6c.png) | ![screen shot 2019-01-31 at 11 23 08 pm](https://user-images.githubusercontent.com/1848368/52108692-9e7ce300-25af-11e9-86c6-98a8412155fe.png) |
| ![screen shot 2019-01-31 at 11 24 29 pm](https://user-images.githubusercontent.com/1848368/52108696-a9d00e80-25af-11e9-8e15-f4b8a38df391.png) | ![screen shot 2019-01-31 at 11 24 17 pm](https://user-images.githubusercontent.com/1848368/52108698-ab013b80-25af-11e9-8e2a-d9379e69fc57.png) |
| ![screen shot 2019-01-31 at 11 36 57 pm](https://user-images.githubusercontent.com/1848368/52109201-77271580-25b1-11e9-8b44-5485c3b49c44.png) | ![screen shot 2019-01-31 at 11 36 49 pm](https://user-images.githubusercontent.com/1848368/52109203-78f0d900-25b1-11e9-8f56-3c5032c59218.png) |

**Comment by Mugen87** [comment]
The original tasks define in https://github.com/mrdoob/three.js/issues/11337#issue-228778022 are implemented via #18108 and #18127.

The issue also discussed the idea to decouple `encoding` into color space and encoding format. If nobody disagrees, I suggest to move this topic to a new issue (if it's still relevant).

**Comment by WestLangley** [comment]
It's still relevant, but not essential yet. I'll take care of it.

It is OK to close this.
